### ISSUE TITLE
Download snapshot is not resilient to connection interruption

### ISSUE THREAD

**Issue by VagueGit** [Mimetis/Dotmim.Sync#470]
I was fortunate to be able to test DMS on a flakey connection. When downloading a 170mb snapshot, there were occasional http exceptions. After each exception DMS stated downloading the snapshot again. It downloaded over 2gb of snapshot before I killed the process.

Our customers are often not as fortunate as us in the quality of their internet connections. We are considering migrating one of our apps from Realm to DMS. DMS would suit this app better than Realm on a reliable connection, but Realm is more resilient to unreliable connections. Where a connection is lost during a large sync, Realm will pickup where it left off.

I experimented with batch sizes from 500 to 10,000, but was unable to get a device to initialise a SQLite database from a SQL Server (over http) using a snapshot, over a flakey connection.

The flakey connection was adequate for web browsing and email. There, just not consistent.

Can DMS be made more resilient to intermittent internet when downloading a snapshot?

**Comment by Mimetis** [comment]
Indeed, that could be a good additional thing to have a kind of resilient http connection.

Actually, `DMS` has already a resilient mechanism to handle connection to databases, I guess we can make the same kind of thing for an http call.

I'm working on it right now, and I think I have found an easy way to implement it, that will break anything from the user perspective.

Jut out of curiosity, how are you handling the resilience / reliable connection with Realm ? Are you using a simple mechanism, some options, or just nothing as it is handled by default ?

**Comment by gentledepp** [comment]
So I guess adding some resiliency for flaky connections should be very easy.
Currently, the `HttpRequestHandler` throws an exception in case the response did not indicate success (`HandleSyncError`):

```csharp
                var args = new HttpSendingRequestMessageArgs(requestMessage, this.orchestrator.GetContext()); 
                await this.orchestrator.InterceptAsync(args, cancellationToken).ConfigureAwait(false);

                // Eventually, send the request
                response = await client.SendAsync(requestMessage, cancellationToken).ConfigureAwait(false);

                if (cancellationToken.IsCancellationRequested)
                    cancellationToken.ThrowIfCancellationRequested();

                // throw exception if response is not successfull
                // get response from server
                if (!response.IsSuccessStatusCode && response.Content != null)
                    await HandleSyncError(response);

```

Now here, you could add resiliency by introducing a [Polly ResiliencePolicy](https://github.com/App-vNext/Polly)

There is a `Retry` policy that should do just that:
https://github.com/App-vNext/Polly#retry
and
https://github.com/App-vNext/Polly/wiki/Retry

There is even a package [Polly.Extensions.Http](https://github.com/App-vNext/Polly.Extensions.Http) which already provides opinionated retry policies for typical http transient errors.
So the solution could very well be to just add that package to `Dotmim.Sync.Web.Client` and 

1. changing the constructor to
```diff

        public HttpRequestHandler(BaseOrchestrator orchestrator)
        {
+            policy = HttpPolicyExtensions
+            .HandleTransientHttpError()
+            .RetryAsync(3); 
            
            this.orchestrator = orchestrator;
        }

``` 

2. and using that policy to wrap the request in
```csharp


                // Eventually, send the request
                response = await this.policy.ExecuteAsync(async () =>
                         await client.SendAsync(requestMessage, cancellationToken).ConfigureAwait(false)
                    ).ConfigureAwait(false);


```

Note: It would be quite nice to also include other transient issues here.
For example: Retry if the hash of the uploaded/downloaded batch does not match with the hash sent - meaning that the data may have been corrupted during upload/download.

In case of a server side error (data corrupted on upload), Dotmim.Sync could return some status code like 4xx or error message indicating this issue. This could then be included in the policy like so:
```diff
            policy = HttpPolicyExtensions
            .HandleTransientHttpError()
+            .OrResult(response => (int)response.StatusCode == someStatusCode) 
            .RetryAsync(3);
```

On the client side (when a corrupted batch was downloaded), Dotmim.Sync could just throw a specific exception like a `DownloadedBatchCorruptedException`
This could then be handled by adding another clause to our policy:


```diff
            policy = HttpPolicyExtensions
            .HandleTransientHttpError()
            .OrResult(response => (int)response.StatusCode == someStatusCode) 
+            .Or<DownloadedBatchCorruptedException>()
            .RetryAsync(3);
```

and  **BOOM** we're done 😎

**Comment by gentledepp** [comment]
Actually, I stumbled across another major improvement we could do: Stream the response data instead of keeping everything in memory.
[This MSDN article](https://docs.microsoft.com/en-us/previous-versions/windows/apps/hh781239(v=win.10)?redirectedfrom=MSDN#create-the-httpclient-send-the-get-request-and-retrieve-the-response) mentions that you should stream responses if the content is larger than 50 MB. 

However, I am sure that on mobile platforms (especially Xamarin.Android), this limit is hit much sooner.
So maybe, @Mimetis , the framework should only read the headers and stream the response into a local file (at least when downloading batches?)

[Here is a quite nice article going over this topic](https://www.tugberkugurlu.com/archive/efficiently-streaming-large-http-responses-with-httpclient)

What do you think? Should I create another improvement issue?

**Comment by VagueGit** [comment]
> Jut out of curiosity, how are you handling the resilience / reliable connection with Realm ? Are you using a simple mechanism, some options, or just nothing as it is handled by default ?

Realm has retry built in.

**Comment by Mimetis** [comment]
**Polly** is a great framework indeed.
Actually, we have a policy class already embedded in `DMS`, that is inspired from **Polly** : [SyncPolicy](https://github.com/Mimetis/Dotmim.Sync/blob/master/Projects/Dotmim.Sync.Core/SyncPolicy.cs).

`DMS` is using a `SyncPolicy` instance for handling connection to the databases.

I'm currently implementing a `SyncPolicy` for the `HttpRequestHandler`.
The idea is to have a default `SyncPolicy` that will handle common issue like session lost of hash mismatch.
You will have the opportunity to pass your own `SyncPolicy` if you want to customize how the retry process is handled.

First tests are pretty good. 
Here is a screenshot of **Fiddler** where my server randomly throw exceptions.

![image](https://user-images.githubusercontent.com/4592555/108698393-60d7b380-7504-11eb-9266-ba5649b22aea.png)

Thanks to the `SyncPolicy` the `WebClientOrchestrator` retries to send the request each time we have an error.

@gentledepp, regarding the stream response, I think we can open a new issue, will be easier to have a conversation ?

**Comment by Mimetis** [comment]
The `SyncPolicy` is integrated by default in the `WebClientOrchestrator` (but you will be able to pass your own policy if you want)

Here is an output on a weird connection :)

``` bash

0%:     Session Begins.
0%:     Getting Server Schema. Scope Name:DefaultScope.
0%:     Provisioned 8 Tables. Provision:Table, TrackingTable, StoredProcedures, Triggers.
10%:    Getting All Changes
42%:    Sending All Changes. Rows:0. Waiting Server Response...
57%:    Getting Batch Changes. (2/10).
59%:    Getting Batch Changes. (3/10).
55%:    Retry Sending Http Request (1/10. Waiting 500ms... )
55%:    Retry Sending Http Request (2/10. Waiting 1000ms... )
61%:    Getting Batch Changes. (4/10).
55%:    Retry Sending Http Request (1/10. Waiting 500ms... )
63%:    Getting Batch Changes. (5/10).
65%:    Getting Batch Changes. (6/10).
55%:    Retry Sending Http Request (1/10. Waiting 500ms... )
67%:    Getting Batch Changes. (7/10).
69%:    Getting Batch Changes. (8/10).
71%:    Getting Batch Changes. (9/10).
55%:    Retry Sending Http Request (1/10. Waiting 500ms... )
55%:    Retry Sending Http Request (2/10. Waiting 1000ms... )
55%:    Retry Sending Http Request (3/10. Waiting 1500ms... )
73%:    Getting Batch Changes. (10/10).
55%:    Retry Sending Http Request (1/10. Waiting 500ms... )
55%:    Retry Sending Http Request (2/10. Waiting 1000ms... )
55%:    Retry Sending Http Request (3/10. Waiting 1500ms... )
55%:    Retry Sending Http Request (4/10. Waiting 2000ms... )
55%:    Retry Sending Http Request (5/10. Waiting 2500ms... )
75%:    [ProductCategory] [Modified] Applied:(41) Total:(41/2752).
76%:    [ProductModel] [Modified] Applied:(128) Total:(169/2752).
76%:    [Product] [Modified] Applied:(48) Total:(217/2752).
77%:    [Product] [Modified] Applied:(77) Total:(246/2752).
77%:    [Product] [Modified] Applied:(110) Total:(279/2752).
77%:    [Product] [Modified] Applied:(149) Total:(318/2752).
78%:    [Product] [Modified] Applied:(204) Total:(373/2752).
78%:    [Product] [Modified] Applied:(253) Total:(422/2752).
79%:    [Product] [Modified] Applied:(285) Total:(454/2752).
79%:    [Product] [Modified] Applied:(295) Total:(464/2752).
83%:    [Address] [Modified] Applied:(450) Total:(914/2752).
86%:    [Customer] [Modified] Applied:(298) Total:(1212/2752).
88%:    [Customer] [Modified] Applied:(596) Total:(1510/2752).
90%:    [Customer] [Modified] Applied:(847) Total:(1761/2752).
94%:    [CustomerAddress] [Modified] Applied:(417) Total:(2178/2752).
95%:    [SalesOrderHeader] [Modified] Applied:(32) Total:(2210/2752).
100%:   [SalesOrderDetail] [Modified] Applied:(542) Total:(2752/2752).
100%:   [Total] Applied:2752. Conflicts:0.
100%:   Session Ended.
Synchronization done.
        Total changes  uploaded: 0
        Total changes  downloaded: 2752
        Total changes  applied: 2752
        Total resolved conflicts: 0
        Total duration :0:0:22.372
```

If you want to play with the last commit on `v0.6.2`, here is a circuit breaker you can implement on your web server to simulate a web exception:

``` cs
[HttpPost]
public async Task Post()
{
    WebServerOrchestrator.TryGetHeaderValue(this.HttpContext.Request.Headers, "dotmim-sync-session-id", out var sessionId);

    var orchestrator = webServerManager.GetOrchestrator(this.HttpContext);

    if (webServerManager.Cache.TryGetValue<SessionCache>(sessionId, out var sessionCache))
    {
        WebServerOrchestrator.TryGetHeaderValue(this.HttpContext.Request.Headers, "dotmim-sync-step", out string iStep);

        var step = (HttpStep)Convert.ToInt32(iStep);

        if (step == HttpStep.GetMoreChanges)
        {
            var random = new Random().Next(100);
            if (random > 50)
                throw new TimeoutException("Error waiting");
        }

    }

    await webServerManager.HandleRequestAsync(this.HttpContext);
}
```

**Comment by Mimetis** [comment]
The *retry on error* will be implemented in `v0.6.2` !
Thanks a lot @VagueGit and @gentledepp for your ideas and suggestions.

The `WebClientOrchestrator` object will use a default `SyncPolicy`, configured with:
- 10 retries for one particular http request before failing.
- 500 ms multiply by current retry count, as a delay before next try.
- If server lost the session for some reason the retry will abort immediately.

If you want to customize the `SyncPolicy` in place, you can access the properties through:
``` cs
var webClientOrchestrator = new WebClientOrchestrator(uri);

// override default 10 max retry count to 20
webClientOrchestrator.SyncPolicy.RetryCount = 20;
// override default duration wait, with a duration = 1sec, then 2sec, then 3sec ...
webClientOrchestrator.SyncPolicy.SleepDuration = iteration => TimeSpan.FromSeconds(1) * iteration;

```

And if you want to go even deeper and having more control on your policy, you can create and provide your own policy.
Here is an example:
``` cs
var policy = SyncPolicy.WaitAndRetry(10,
// Gets the duration to wait before next retry
(retryNumber) =>
{
    // Sliding duration of 500, 1000, 1500, 2000 ms etc....
    return TimeSpan.FromMilliseconds(500 * retryNumber);
},    
// Gets a boolean indicating if we can retry of if we should stop and rethrow the exception
(ex, arg) =>
{
    var webEx = ex as SyncException;

    // handle session lost. if it's a session lost, just immediately fail
    return webEx == null || webEx.TypeName != nameof(HttpSessionLostException);
}
// Gets an Interceptor that allows you to do something during each retry
, async (ex, cpt, ts, arg) =>
{
   Console.WriteLine($"Try failed. Attempt number {cpt}"); 
});

var webClientOrchestrator = new WebClientOrchestrator(uri, syncPolicy=policy);

```
